### ISSUE TITLE
Not saving session if headers were sent

### ISSUE THREAD

**Issue by djfd** [kohana/kohana#101]
Is it correct?
E.g. consider the situation of generated downloadable content, need to be different for every request during the session, further sent via Response's send_file(), like this:

```
$seq_no = Session::instance()->get('seq-no', 0);
$this->response->body(produce(++$seq_no));
Session::instance()->set('seq-no', $seq_no);
$this->response->send_file(TRUE, 'sequencer.json', array('inline'=>TRUE));
```

this wont work because when Session's save method is  called the headers were already sent:
system/classes/Kohana/Session.php:361

```
        if (headers_sent() OR $this->_destroyed)
        {
            // Session cannot be written when the headers are sent or when
            // the session has been destroyed
            return FALSE;
        }
```

**Comment by acoulton** [comment]
Yes, that looks like a bug to me. The `headers_sent()` check should only be relevant for the Cookie session driver (and any third-party drivers that set cookies). For native, database and other server-side sessions the session should still be saved.

Could you submit a PR?

**Comment by enov** [comment]
@acoulton I guess all drivers send the session ID via cookie. No?

I remember there was a discussion about it somewhere in the old forums:

http://forum.kohanaframework.org/discussion/12041/problem-with-sessions/p1

**Comment by djfd** [comment]
I am sorry, what is PR?

and. ofc, not all sessions are headers based. E.g. one can implement session wake-up using $_GET variable

**Comment by acoulton** [comment]
@enov the session ID cookie is set in `session_start` by PHP native sessions, the `session_write_close` operation is entirely server-side. Kohana's database driver **should** (no idea if it is) be the same.

@djfd PR = Pull request for a bug fix.

**Comment by djfd** [comment]
hmm... interesting...

that is all that we wrote here is not enough? need to emit PR as well?

**Comment by djfd** [comment]
And, anyway,

Session is session, headers are absolutely irrelevant here... Sessions can be instantiated using a lot of means (imaginate SMS sessions, do you can?) 

So why headers here?

**Comment by enov** [comment]
@acoulton @djfd both the native and the database drivers use cookies (which are basically headers) to sync the session id with the server:

https://github.com/kohana/core/blob/3.3/master/classes/Kohana/Session/Native.php#L45

https://github.com/kohana/database/blob/3.3/master/classes/Kohana/Session/Database.php#L174

**Comment by djfd** [comment]
@enov
yes, they do. so possible it is their particular isuue if they will send headers or not. why generic session does handle that?

**Comment by enov** [comment]
This is by design. This is not a bug.

**Comment by djfd** [comment]
:D nobody tells that is a kind of  bug

just wrong design

**Comment by acoulton** [comment]
@djfd what you have here is a bug report. It's enough to identify the problem, but not enough to fix it.

To fix it, someone needs to implement tests, review the other drivers and move the `headers_sent` check to them if they send cookies (eg the cookie session driver) and implement tests for those.

Since everyone is a volunteer and pressed for time, yes - we'd appreciate it if you could create a pull request with the full required fix.

@enov disagree. This is a bug. The native session sends a cookie on `_read`, not on `_write`. There is absolutely no valid reason to skip the `_write` call if headers are sent. The database session could also be perfectly easily refactored to set the ID (and cookie) at the start rather than the end of the request. @djfd is quite right that it should be the drivers that check for `headers_sent` if it matters to them, not the generic base class.

**Comment by enov** [comment]
> just wrong design

No problem :D

> The native session sends a cookie on _read

I don't think it sends a cookie on `_read`, it just _sets_ the session id. Headers are sent just before the body is sent, in [bootstrap.php](https://github.com/kohana/kohana/blob/3.3/master/index.php#L119)

**Comment by djfd** [comment]
@enov 
please look at Response::send_file()
send header, then send body, flush, how can I save my SMS session established via satellite channel?

(yeah, a kind of joke, but really, who did tell that all sessions are cookie based?)

**Comment by acoulton** [comment]
@enov almost :)

Headers are only sent there if the request completes normally. If you call `response->send_file()` then headers are sent [from that method](https://github.com/kohana/core/blob/3.3/master/classes/Kohana/Response.php#L504) which finishes with exit.

However, that's also misleading because both the native `session_start` and Kohana's `Cookie::set` use PHP's native `setcookie`. PHP sends those headers immediately before the first byte of output, wherever that comes from and regardless of what userland code says. `request->send_headers()` isn't relevant in this context.

So yes, technically speaking the cookie isn't _sent_ on `_read` but it is queued and guaranteed to be sent between that call and you sending any output to the client.

**Comment by enov** [comment]
Whoa! I never knew such a `send_file` method existed :smile: 

I usually do a standard `$response->body()` after setting the right `content-type` header.

But, that method has a good warning doc `[!!] No further processing can be done after this method is called!`

**Comment by acoulton** [comment]
@enov thing is this isn't about processing after the file is sent. Values set in the session **before** the file is sent will not currently be saved because the session shutdown comes after the headers are sent and therefore the generic Session class never calls the `_write` method on the driver.

Although, I've just looked again and in fact this bug doesn't affect the native driver because it uses a reference to $_SESSION as its data storage and PHP will persist that with or without a call to session_write_close.

So actually I think the issue only applies to:
- cookie sessions (but by definition this fix doesn't work because you can't send a cookie after headers are sent)
- database sessions (but again the current implementation, probably incorrectly, sets a cookie during _write)
- third-party session-via-sms-and-satellite-drivers

The `headers_sent` check should still be moved to relevant drivers, but possibly also there needs to be a call to trigger a session save from `send_file`.

@djfd what session driver are you using?

**Comment by djfd** [comment]
@acoulton
I do use Db Sessions triggered up by super-cookie (for now it uses cookies but I have an idea of moving it to either get or post array). super cookie it is a piece of encrypted information including session id and some specific variables. This looks as a cookie session but, internally handled information volume is too big to be pure cookie-session, thus this mixed approach

**Comment by acoulton** [comment]
@djfd is that a custom db session driver or a kohana one?

**Comment by djfd** [comment]
@acoulton
pure kohana. with a minor overrides

**Comment by acoulton** [comment]
@djfd ok then as @enov says your fix here isn't going to be the whole solution, the database driver also needs to be refactored so that the session id cookie is set at the time it's generated rather than during `_write`. Otherwise removing the `headers_sent` check will just go on to trigger an error from the [`Cookie::set()` call in the database driver](https://github.com/kohana/database/blob/3.3/master/classes/Kohana/Session/Database.php#L174)

**Comment by djfd** [comment]
hmm, but Cookie::set unnecessarily need to call PHP's setcookie(), it can instead accumulate them somewhere and then send when actual output begins?

**Comment by djfd** [comment]
As I see it
container.start()
...
application activity (output, headers, anything)
...
<PHP.END (exit, die, normal finish)>
container.finish() { // triggered by any script termination
send headers
send content
}

**Comment by acoulton** [comment]
@djfd it's much safer (and less work) to let PHP's setcookie accumulate the cookies and send them when output begins which is what currently happens.

The issue you've found is that the database driver is trying to send a cookie **after** output begins - and that will never work.

Normal requests work inside a container much like you have there, but you generally don't want that for sending a file because you'd then be buffering the file in memory which is almost always a bad idea.  You could trigger a session write just before the file is sent, but there's no guarantee a session has been started or which driver(s) need to be written so that would be problematic.

**Comment by djfd** [comment]
well, it is exactly what I am doing right now: check if session was started then save it, after that send my just a few KB's file (regarding the bad idea of making files in the memory)))

yeah, it is better if all work is done by means of system software, but for current session save() triggering method it will be 99% beyond the cookies send event. So I just cannot see [right now] any alternative to some supervisor interface handling that (I mean session events: init, destroy, regenerate, rename etc) and taking appropriate actions at right time

**Comment by djfd** [comment]
or, if PHP engine is smart enough to replace the same cookie that is being set multiple times, we can set the cookies right after we feeling that right: session ID changed - ok, send cookie right now, destroyed - delete cookie right now, etc...

well it is possible, end user will receive wrong ID of destroyed session. Is it issue? Surely nope, session will be regenerated upon next request, new ID sent, etc... I only can see this the only negative impact of asynchronous cookies sending [presuming enough PHP smartness, need to check how it does behave actually]

**Comment by djfd** [comment]
and main point was that generic abstract session is absolutely unrelated to any headers, it is implementors/extenders task to select and use proper ID reporting schema

**Comment by djfd** [comment]
@acoulton 
Hi,
maybe you will be interested. I almost completed structured sessions, they are for Phalcon (https://phalconphp.com/en/) but actually do use the only specific thing from it (EventsManager), which can be easily stripped down. Other dependencies are quite obvious and can be replaced with their kohana counterparts (DB, Cookies). So if you want I can send you interface files to look, to think about, possible to make some critics.

Later on, if approach is ok I can send full files set so you can adopt it to kohana.

A few highlights: multiple simultaneous sessions are possible (including single PHP native session, if required) - this can be used for example by site admin to log in into website as a regular customer, while keeping self logged in into backend, using parallel session for that; highly structured and split into separate functional parts allowing independent configuration; resolves that we were discussing here; an arbitrary session ID carrier; an arbitrary encoder; an arbitrary persistence; supporting three main modes of operation (read-only, concurrent, locking) right out of the box ))

Thanks
